lms/djangoapps/courseware/module_render.py

```diff
@@ -462,17 +474,16 @@ def inner_get_module(descriptor):
         )
 
     def publish(block, event_type, event):
-        """A function that allows XModules to publish events."""
-        if event_type == 'grade' and not is_masquerading_as_specific_student(user, course_id):
-            SCORE_PUBLISHED.send(
-                sender=None,
-                block=block,
-                user=user,
-                raw_earned=event['value'],
-                raw_possible=event['max_value'],
-                only_if_higher=event.get('only_if_higher'),
-                score_deleted=event.get('score_deleted'),
-            )
+        """
+        A function that allows XModules to publish events.
+        """
+        special_events = {
+            'completion': handle_completion_event,
+            'grade': handle_grade_event,
+            'progress': handle_progress_event,
+        }
```

Comment:
Well, this isn't actually a class, so publish isn't technically a method.  It's a weird jumbo function that declares a bunch of inner functions (closures, technically, since they depend on variables declared within the outer function.  So I can create another inner function to pick the appropriate special handler, but I can't make it, for instance, a class variable.

Do you think the else branch should not be an else, but just the thing that happens unconditionally after the special handling?  That seems sensible, but it's a change from how grades were being handled.  @robrap, do you have an opinion on this?

Reply:
Yep, I get it. I was just suggesting you move this local dictionary declaration up one level (still inside the outer function, but not inside this particular inner function) - it would still be in scope to be able to reference the closures like `handle_grade_event`. It's not a big deal though.

> Do you think the else branch should not be an else, but just the thing that happens unconditionally after the special handling?

Yep, that would make more sense to me. I just don't know the reason why it wasn't the case for grades.

Reply:
(I'm not sure I understand what the asides are for, so I can't judge rightly whether all that extra event context processing should be performed on grades and completions, except to the extent that I don't see any reason why they should be different)

Reply:
Making that change caused several test errors, so I'm going to leave it as an either/or thing.